src/sfctl/custom_cluster.py

```diff
@@ -47,32 +53,84 @@ def select(endpoint, cert=None, key=None, pem=None, ca=None, no_verify=False):
-                              set_no_verify)
+    from sfctl.config import (set_ca_cert, set_auth,
+                              set_cluster_endpoint, client_endpoint,
+                              set_no_verify, no_verify_setting)
     from msrest import ServiceClient, Configuration
-    from sfctl.auth import ClientCertAuthentication
+    from sfctl.auth import ClientCertAuthentication, AdalAuthentication
 
-    select_arg_verify(endpoint, cert, key, pem, ca, no_verify)
+    select_arg_verify(endpoint, cert, key, pem, ca, aad, no_verify)
 
-    client_cert = None
-    if pem:
-        client_cert = pem
-    elif cert:
-        client_cert = (cert, key)
+    access_token = None
+    if aad:
+        endpoint_backup = client_endpoint()
+        noverify_backup = no_verify_setting()
+        set_cluster_endpoint(endpoint)
+        set_no_verify(no_verify)
 
-    rest_client = ServiceClient(
-        ClientCertAuthentication(client_cert, ca, no_verify),
-        Configuration(endpoint)
-    )
+        access_token = sf_get_aad_token()
 
-    # Make sure basic GET request succeeds
-    rest_client.send(rest_client.get('/')).raise_for_status()
+        set_cluster_endpoint(endpoint_backup)
+        set_no_verify(noverify_backup)
+
+        rest_client = ServiceClient(
+            AdalAuthentication(access_token, no_verify),
+            Configuration(endpoint)
+        )
+
+        # Make sure basic GET request succeeds
+        rest_client.send(rest_client.get('/')).raise_for_status()
+    else:
+        client_cert = None
+        if pem:
+            client_cert = pem
+        elif cert:
+            client_cert = (cert, key)
+
+        rest_client = ServiceClient(
+            ClientCertAuthentication(client_cert, ca, no_verify),
+            Configuration(endpoint)
+        )
+
+        # Make sure basic GET request succeeds
+        rest_client.send(rest_client.get('/')).raise_for_status()
 
-    set_ca_cert(ca)
-    set_cert(pem, cert, key)
-    set_no_verify(no_verify)
     set_cluster_endpoint(endpoint)
+    set_no_verify(no_verify)
+    set_ca_cert(ca)
+    set_auth(pem, cert, key, access_token)
+
+def sf_get_aad_token():
+    """Get AAD token"""
+    from sfctl.config import set_config_value
+    from sfctl.apiclient import create
+    from datetime import datetime
+    from datetime import timedelta
+
+    set_config_value('security', 'none')
+    sf_client = create(None)
+    aad_metadata = sf_client.get_aad_metadata()
+
+    if aad_metadata.type != "aad":
+        raise CLIError("Not AAD cluster")
+
+    aad_resource = aad_metadata.metadata
+
+    tenant_id = aad_resource.tenant
+    context = adal.AuthenticationContext(aad_resource.login + '/' + tenant_id,
+                                         api_version=None)
+    cluster_id = aad_resource.cluster
+    client_id = aad_resource.client
+
+    code = context.acquire_user_code(cluster_id, client_id)
+    print(code['message'])
```

Comment:
This is not log, it is to print AAD auth device code for user to authenticate on web browser.